### PR TITLE
code review

### PR DESCRIPTION
**Pull request by wrighbr**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,99 @@
-# backstage-plugin-gkeusage
\ No newline at end of file
+# Backstage GKE Usage plugin
```
Comment (Ticalie) [inline README.md:1]: Is this file autogenerated?

File: rollup.config.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
+/* eslint-disable import/no-extraneous-dependencies */
```
Comment (Ticalie) [inline rollup.config.js:1]: Remove the disable :)

File: src/api/index.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  description: "Used by the GKEUsage plugin to make requests",
+});
+
+export type Options = {
+  discoveryApi: DiscoveryApi;
+  proxyPath?: string;
+};
+
+export class GkeusageApiClient implements gkeusageApi {
+  // @ts-ignore
```
Comment (Ticalie) [inline src/api/index.ts:16]: Why ignore?

File: src/components/GKEUsageConsumption.tsx [hunk 4]
```diff
@@ -0,0 +1,182 @@
+import { Grid } from "@material-ui/core";
```
Comment (Ticalie) [inline src/components/GKEUsageConsumption.tsx:1]: I would build this as a functional component and use hooks :)

File: src/components/GKEUsageConsumption.tsx [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+      `${this.props.url}&maxAge=${this.props.maxAge}`
+    ).then((res) => res.json());
+
+    let cpuUsage = 0;
+    let memoryUsage = 0;
+    let storageUsage = 0;
+    let networkUsage = 0;
+    let gpuUsage = 0;
+
+    response.forEach(function setPercentage(value: any) {
```
Comment (Ticalie) [inline src/components/GKEUsageConsumption.tsx:31]: What is reponse is null?

File: src/components/GKEUsageConsumption.tsx [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+      networkUsage,
+      gpuUsage,
+      maxAge,
+      isLoaded: true,
+    });
+  }
+
+  async componentDidUpdate(prevState: { maxAge: string }) {
+    if (prevState.maxAge !== this.props.maxAge) {
+      /* eslint-disable */
```
Comment (Ticalie) [inline src/components/GKEUsageConsumption.tsx:68]: Why disable? :)

File: src/components/GKEUsageConsumption.tsx [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+        `${this.props.url}&maxAge=${this.props.maxAge}`
+      ).then((res) => res.json());
+
+      let cpuUsage = 0;
+      let memoryUsage = 0;
+      let storageUsage = 0;
+      let networkUsage = 0;
+      let gpuUsage = 0;
+
+      response.forEach(function setPercentage(value: any) {
```
Comment (Ticalie) [inline src/components/GKEUsageConsumption.tsx:84]: What if response is null?

File: src/components/GKEUsageConsumption.tsx [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+          case "gpu":
+            gpuUsage = value.consumption_percentage;
+            break;
+          default:
+            break;
+        }
+      });
+
+      const maxAge = this.props.maxAge;
+      /* eslint-disable */
```
Comment (Ticalie) [inline src/components/GKEUsageConsumption.tsx:107]: Why disable?

File: src/components/GKEUsageCost.tsx [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+    const currency: any = getSymbolFromCurrency(response[0].currency);
+
+    let memoryCost = 0;
+    let cpuCost = 0;
+    let networkCost = 0;
+    let storageCost = 0;
+    let gpuCost = 0;
+
+    response.forEach(function roundNumbers(value: any) {
```
Comment (Ticalie) [inline src/components/GKEUsageCost.tsx:46]: null check?

File: src/components/GKEUsageCost.tsx [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+    if (response.hasOwnProperty("error")) {
+      this.setState({
+        error: true,
+        errorMsg: response.error.message,
+        isLoaded: true,
+      });
+    }
+
+    const currency: any = getSymbolFromCurrency(response[0].currency);
```
Comment (Ticalie) [inline src/components/GKEUsageCost.tsx:38]: If you use `any` you're pretty much saying "I don't want any type-safety"

File: src/components/GKEUsageCost.tsx [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+      totalCost,
+      currency,
+      maxAge,
+      isLoaded: true,
+    });
+  }
+
+  async componentDidUpdate(prevState: { maxAge: string }) {
+    if (prevState.maxAge !== this.props.maxAge) {
+      /* eslint-disable */
```
Comment (Ticalie) [inline src/components/GKEUsageCost.tsx:90]: disable?

File: src/components/GKEUsageCost.tsx [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+      const currency: any = getSymbolFromCurrency(response[0].currency);
+
+      let memoryCost = 0;
+      let cpuCost = 0;
+      let networkCost = 0;
+      let storageCost = 0;
+      let gpuCost = 0;
+      let totalCost = 0;
+
+      response.forEach(function roundNumbers(value: any) {
```
Comment (Ticalie) [inline src/components/GKEUsageCost.tsx:108]: null check?

File: src/components/GKEUsageCost.tsx [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+            break;
+          default:
+            break;
+        }
+      });
+
+      totalCost = memoryCost + cpuCost + networkCost + storageCost + gpuCost;
+
+      const maxAge = this.props.maxAge;
+      /* eslint-disable */
```
Comment (Ticalie) [inline src/components/GKEUsageCost.tsx:133]: disable?

File: src/components/GKEUsageDashboardPage.tsx [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+      <Select defaultValue={days} onChange={onchangeSelect} options={options} />
+      <Card>
+        <GKECost url={costUrl} maxAge={days.value} />
+      </Card>
+      <p />
+      <Card>
+        <p />
+        <GKEConsumption url={usageUrl} maxAge={days.value} />
+      </Card>
+      {/* <p></p>
```
Comment (Ticalie) [inline src/components/GKEUsageDashboardPage.tsx:62]: remove this

File: src/components/GKEUsageDashboardPage.tsx [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+ */
+
+import { Entity } from "@backstage/catalog-model";
+import { InfoCard } from "@backstage/core";
+import React, { useState } from "react";
+import Select from "react-select";
+import { useGkeUsageMeteringAppData } from "./useGkeUsageMeteringAppData";
+import { GKECost } from "./GKEUsageCost";
+import { GKEConsumption } from "./GKEUsageConsumption";
+// import { GKEEfficiency } from './GKEUsageEfficiency'
```
Comment (Ticalie) [inline src/components/GKEUsageDashboardPage.tsx:24]: remove this

File: src/components/GKEUsageEfficiency.tsx [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+    let storageUsage = 0;
+    let networkUsage = 0;
+    let gpuUsage = 0;
+    let memoryCost = 0;
+    let cpuCost = 0;
+    let networkCost = 0;
+    let storageCost = 0;
+    let gpuCost = 0;
+
+    usage.forEach(function setPercentage(value: any) {
```
Comment (Ticalie) [inline src/components/GKEUsageEfficiency.tsx:33]: null check?

File: src/components/GKEUsageEfficiency.tsx [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+          break;
+        case "gpu":
+          gpuUsage = value.consumption_percentage;
+          break;
+        default:
+          break;
+      }
+    });
+
+    cost.forEach(function roundNumbers(value: any) {
```
Comment (Ticalie) [inline src/components/GKEUsageEfficiency.tsx:55]: null check?

File: src/components/GKEUsageEfficiency.tsx [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+      gpuEfficiency,
+    } = this.state;
+    return (
+      <div>
+        <Grid container spacing={5}>
+          <Grid item direction="row" alignItems="center">
+            <ReactSpeedometer
+              value={cpuEfficiency}
+              maxValue={100}
+              // currentValueText={'cpu efficiency'}
```
Comment (Ticalie) [inline src/components/GKEUsageEfficiency.tsx:107]: remove this

File: src/components/GKEUsageEfficiency.tsx [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+            />
+          </Grid>
+
+          <Grid item direction="row" alignItems="center">
+            <ReactSpeedometer
+              value={memoryEfficiency}
+              maxValue={100}
+              // currentValueText={'memory efficiency'}
```
Comment (Ticalie) [inline src/components/GKEUsageEfficiency.tsx:115]: remove this

File: src/components/GKEUsageEfficiency.tsx [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+            />
+          </Grid>
+
+          <Grid item direction="row" alignItems="center">
+            <ReactSpeedometer
+              value={networkEfficiency}
+              maxValue={100}
+              // currentValueText={'network efficiency'}
```
Comment (Ticalie) [inline src/components/GKEUsageEfficiency.tsx:123]: remove this

File: src/components/GKEUsageEfficiency.tsx [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+            />
+          </Grid>
+
+          <Grid item direction="row" alignItems="center">
+            <ReactSpeedometer
+              value={storageEfficiency}
+              maxValue={100}
+              // currentValueText={'storage efficiency'}
```
Comment (Ticalie) [inline src/components/GKEUsageEfficiency.tsx:131]: remove this

File: src/components/GKEUsageEfficiency.tsx [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+            />
+          </Grid>
+
+          <Grid item direction="row" alignItems="center">
+            <ReactSpeedometer
+              value={gpuEfficiency}
+              maxValue={100}
+              // currentValueText={'gpu efficiency'}
```
Comment (Ticalie) [inline src/components/GKEUsageEfficiency.tsx:139]: remove this

File: src/components/data.ts [hunk 23]
```diff
@@ -0,0 +1,49 @@
+export const options = [
```
Comment (Ticalie) [inline src/components/data.ts:1]:
```suggestion
interface Option: { label: string,
    value: number}

export const options: Option[] = [
```

File: src/components/data.ts [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+    label: "last 30 days",
+    value: "30",
+  },
+  {
+    label: "last 60 days",
+    value: "60",
+  },
+];
+
+export const columns = [
```
Comment (Ticalie) [inline src/components/data.ts:24]:
```suggestion
interface Column = {name: string,
    selector: string}

export const columns: Column[] = [
```

File: src/plugin.ts [hunk 25]
```diff
@@ -0,0 +1,39 @@
+import { Entity } from "@backstage/catalog-model";
+import {
+  createApiFactory,
+  // createComponentExtension,
```
Comment (Ticalie) [inline src/plugin.ts:4]: remove this

**Review by Ticalie — COMMENTED** [review]
Overall good looking.
I would consider using functional components and hooks.
I would move all my HTTP calls into a .service.ts file

File: src/components/GKEUsageConsumption.tsx (hunk 4 above)
Comment (wrighbr) [reply src/components/GKEUsageConsumption.tsx:2]: Have converted to  functional component and use hooks

File: src/components/GKEUsageDashboardPage.tsx (hunk 15 above)
Comment (wrighbr) [reply src/components/GKEUsageDashboardPage.tsx]: done

File: src/components/GKEUsageDashboardPage.tsx (hunk 14 above)
Comment (wrighbr) [reply src/components/GKEUsageDashboardPage.tsx]: done

File: src/components/GKEUsageEfficiency.tsx (hunk 16 above)
Comment (wrighbr) [reply src/components/GKEUsageEfficiency.tsx:33]: Component not used so I've removed it

File: src/components/GKEUsageEfficiency.tsx (hunk 17 above)
Comment (wrighbr) [reply src/components/GKEUsageEfficiency.tsx:55]: Component used so I've removed it

File: src/components/GKEUsageEfficiency.tsx (hunk 18 above)
Comment (wrighbr) [reply src/components/GKEUsageEfficiency.tsx:107]: I've refactored this

File: src/components/GKEUsageEfficiency.tsx (hunk 19 above)
Comment (wrighbr) [reply src/components/GKEUsageEfficiency.tsx:115]: I've refactored this

File: src/components/GKEUsageEfficiency.tsx (hunk 20 above)
Comment (wrighbr) [reply src/components/GKEUsageEfficiency.tsx:123]: I've refactored this

File: src/components/GKEUsageEfficiency.tsx (hunk 21 above)
Comment (wrighbr) [reply src/components/GKEUsageEfficiency.tsx:131]: I've refactored this

File: src/components/GKEUsageEfficiency.tsx (hunk 22 above)
Comment (wrighbr) [reply src/components/GKEUsageEfficiency.tsx:139]: I've refactored this

File: src/plugin.ts (hunk 25 above)
Comment (wrighbr) [reply src/plugin.ts:4]: removed

**Pull request merged**